File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/agent/Agent.java
package org.hyperledger.indy.sdk.agent;

import java.util.concurrent.CompletableFuture;

import com.sun.jna.Pointer;
import org.hyperledger.indy.sdk.IndyException;
import org.hyperledger.indy.sdk.IndyJava;
import org.hyperledger.indy.sdk.LibIndy;
import org.hyperledger.indy.sdk.ParamGuard;
import org.hyperledger.indy.sdk.wallet.Wallet;
import org.hyperledger.indy.sdk.agent.AgentResults.ParseMsgResult;

import com.sun.jna.Callback;

/**
 * agent.rs API
 */

/**
 * Agent related functionality.
 */
public class Agent extends IndyJava.API {

	private Agent() {

	}

	/*
	 * STATIC CALLBACKS
	 */

	/**
	 * Callback used when prepMsg completes.
	 */
	private static Callback prepMsgCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err, Pointer encrypted_msg, int encrypted_len) {

			CompletableFuture<byte[]> future = (CompletableFuture<byte[]>) removeFuture(xcommand_handle);
			if (! checkCallback(future, err)) return;

			byte[] result = new byte[encrypted_len];
			encrypted_msg.read(0, result, 0, encrypted_len);
			future.complete(result);
		}
	};

	/**
	 * Callback used when prepAnonymousMsg completes.
	 */
	private static Callback prepAnonymousMsgCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err, Pointer encrypted_msg, int encrypted_len) {

			CompletableFuture<byte[]> future = (CompletableFuture<byte[]>) removeFuture(xcommand_handle);
			if (! checkCallback(future, err)) return;

			byte[] result = new byte[encrypted_len];
			encrypted_msg.read(0, result, 0, encrypted_len);
			future.complete(result);
		}
	};

	/**
	 * Callback used when parseMsg completes.
	 */
	private static Callback parseMsgCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err, String senderKey, Pointer msg_data, int msg_len) {

			byte[] msg = new byte[msg_len];
			msg_data.read(0, msg, 0, msg_len);

			CompletableFuture<ParseMsgResult> future = (CompletableFuture<ParseMsgResult>) removeFuture(xcommand_handle);
			if (! checkCallback(future, err)) return;

			ParseMsgResult result = new ParseMsgResult(senderKey, msg);
			future.complete(result);


		}
	};

	/*
	 * STATIC METHODS
	 */

	/**
	 * @param wallet       The wallet.
	 * @param senderKey
	 * @param recipientKey
	 * @param message      a message to be encrypted
	 * @return A future that resolves to an encrypted message.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<byte[]> prepMsg(
			Wallet wallet,
			String senderKey,
			String recipientKey,
			byte[] message) throws IndyException {

		ParamGuard.notNull(wallet, "wallet");
		ParamGuard.notNullOrWhiteSpace(senderKey, "senderKey");
		ParamGuard.notNullOrWhiteSpace(recipientKey, "recipientKey");
		ParamGuard.notNull(message, "message");

		CompletableFuture<byte[]> future = new CompletableFuture<byte[]>();
		int commandHandle = addFuture(future);

		int walletHandle = wallet.getWalletHandle();

		int result = LibIndy.api.indy_prep_msg(
				commandHandle,
				walletHandle,
				senderKey,
				recipientKey,
				message,
				message.length,
				prepMsgCb);

		checkResult(result);

		return future;
	}

	/**
	 * @param recipientKey
	 * @param message      a message to be encrypted
	 * @return A future that resolves to an encrypted message.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<byte[]> prepAnonymousMsg(
			String recipientKey,
			byte[] message) throws IndyException {

		ParamGuard.notNullOrWhiteSpace(recipientKey, "recipientKey");
		ParamGuard.notNull(message, "message");

		CompletableFuture<byte[]> future = new CompletableFuture<byte[]>();
		int commandHandle = addFuture(future);

		int result = LibIndy.api.indy_prep_anonymous_msg(
				commandHandle,
				recipientKey,
				message,
				message.length,
				prepMsgCb);

		checkResult(result);

		return future;
	}

	/**
	 * @param wallet       The wallet.
	 * @param recipientKey
	 * @param encryptedMsg encrypted message
	 * @return A future that resolves to senderKey and message.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<ParseMsgResult> parseMsg(
			Wallet wallet,
			String recipientKey,
			byte[] encryptedMsg) throws IndyException {

		ParamGuard.notNull(wallet, "wallet");
		ParamGuard.notNullOrWhiteSpace(recipientKey, "recipientKey");
		ParamGuard.notNull(encryptedMsg, "encryptedMsg");

		CompletableFuture<ParseMsgResult> future = new CompletableFuture<ParseMsgResult>();
		int commandHandle = addFuture(future);

		int walletHandle = wallet.getWalletHandle();

		int result = LibIndy.api.indy_parse_msg(
				commandHandle,
				walletHandle,
				recipientKey,
				encryptedMsg,
				encryptedMsg.length,
				parseMsgCb);

		checkResult(result);

		return future;
	}
}

File: wrappers/dotnet/indy-sdk-dotnet/AgentApi/Agent.cs
using Hyperledger.Indy.Utils;
using Hyperledger.Indy.WalletApi;
using System.Runtime.InteropServices;
using System.Threading.Tasks;
using static Hyperledger.Indy.AgentApi.NativeMethods;

namespace Hyperledger.Indy.AgentApi
{
    /// <summary>
    /// Provides functionality to enable communication between agents.
    /// </summary>
    public static class Agent
    {
        /// <summary>
        /// Gets the callback to use when commands that prepare messages have completed.
        /// </summary>
        private static PrepareMsgCompletedDelegate _agentMessagePreparedCallback = (xcommand_handle, err, encrypted_data, encrypted_len) =>
        {
            var taskCompletionSource = PendingCommands.Remove<byte[]>(xcommand_handle);

            if (!CallbackHelper.CheckCallback(taskCompletionSource, err))
                return;

            var bytes = new byte[encrypted_len];
            Marshal.Copy(encrypted_data, bytes, 0, encrypted_len);

            taskCompletionSource.SetResult(bytes);
        };

        /// <summary>
        /// Gets the callback to use when the ParsMsgAsync command has completed.
        /// </summary>
        private static ParseMsgCompletedDelegate _agentMessageParsedCallback = (xcommand_handle, err, sender_key, msg_data, msg_len) =>
        {
            var taskCompletionSource = PendingCommands.Remove<ParseMsgResult>(xcommand_handle);

            if (!CallbackHelper.CheckCallback(taskCompletionSource, err))
                return;

            var bytes = new byte[msg_len];
            Marshal.Copy(msg_data, bytes, 0, msg_len);

            var result = new ParseMsgResult(sender_key, bytes);

            taskCompletionSource.SetResult(result);
        };

        /// <summary>
        /// Prepares a message so that it can be securely sent to a recipient.
        /// </summary>
        /// <param name="wallet">The wallet containing the keys of the sender and the recipient.</param>
        /// <param name="senderKey">The public key of the sender.</param>
        /// <param name="recipientKey">The verification key of the intended recipient of the message.</param>
        /// <param name="message">The message content to prepare.</param>
        /// <returns>An asynchronous <see cref="Task{T}"/> that resolves to an array of bytes containing the prepared message.</returns>
        /// <exception cref="WalletValueNotFoundException">Thrown if the sender key does not exist in the <paramref name="wallet"/>.</exception>
        /// <exception cref="InvalidStructureException">Thrown if the <paramref name="recipientKey"/> is invalid.</exception>
        public static Task<byte[]> PrepMsgAsync(Wallet wallet, string senderKey, string recipientKey, byte[] message)
        {
            ParamGuard.NotNull(wallet, "wallet");
            ParamGuard.NotNullOrWhiteSpace(senderKey, "senderKey");
            ParamGuard.NotNullOrWhiteSpace(recipientKey, "recipientKey");
            ParamGuard.NotNull(message, "message");

            var taskCompletionSource = new TaskCompletionSource<byte[]>();
            var commandHandle = PendingCommands.Add(taskCompletionSource);

            var result = NativeMethods.indy_prep_msg(
                commandHandle,
                wallet.Handle,
                senderKey,
                recipientKey,
                message,
                message.Length,
                _agentMessagePreparedCallback);

            CallbackHelper.CheckResult(result);

            return taskCompletionSource.Task;
        }

        /// <summary>
        /// Prepares an anonymous message so that it can be securely sent to a recipient.
        /// </summary>
        /// <param name="recipientKey">The verification key of the intended recipient of the message.</param>
        /// <param name="message">The message content to prepare.</param>
        /// <returns>An asynchronous <see cref="Task{T}"/> that resolves to an array of bytes containing the prepared message.</returns>
        /// <exception cref="InvalidStructureException">Thrown if the value passed to the <paramref name="recipientKey"/> is malformed or the content of the <paramref name="message"/> parameter is invalid.</exception>
        public static Task<byte[]> PrepAnonymousMsgAsync(string recipientKey, byte[] message)
        {
            ParamGuard.NotNullOrWhiteSpace(recipientKey, "recipientKey");
            ParamGuard.NotNull(message, "message");

            var taskCompletionSource = new TaskCompletionSource<byte[]>();
            var commandHandle = PendingCommands.Add(taskCompletionSource);

            var result = NativeMethods.indy_prep_anonymous_msg(
                commandHandle,
                recipientKey,
                message,
                message.Length,
                _agentMessagePreparedCallback);

            CallbackHelper.CheckResult(result);

            return taskCompletionSource.Task;
        }

        /// <summary>
        /// Parses an encrypted message prepared by another party.
        /// </summary>
        /// <param name="wallet">The wallet containing the keys.</param>
        /// <param name="recipientKey">The verification key of the recipient.</param>
        /// <param name="encryptedMsg">The encrypted message to parse.</param>
        /// <returns>An asynchronous <see cref="Task{T}"/> that resolves to a <see cref="ParseMsgResult"/> containing the sender key and the parsed message 
        /// content.</returns>
        /// <exception cref="InvalidStructureException">Thrown if the value passed to the <paramref name="recipientKey"/> parameter is malformed or the <paramref name="encryptedMsg"/> is invalid.</exception>
        /// <exception cref="WalletValueNotFoundException">Thrown if the <paramref name="recipientKey"/> does not exist in the <paramref name="wallet"/>.</exception>
        public static Task<ParseMsgResult> ParseMsgAsync(Wallet wallet, string recipientKey, byte[] encryptedMsg)
        {
            ParamGuard.NotNull(wallet, "wallet");
            ParamGuard.NotNullOrWhiteSpace(recipientKey, "recipientKey");
            ParamGuard.NotNull(encryptedMsg, "encryptedMsg");

            var taskCompletionSource = new TaskCompletionSource<ParseMsgResult>();
            var commandHandle = PendingCommands.Add(taskCompletionSource);

            var result = NativeMethods.indy_parse_msg(
                commandHandle,
                wallet.Handle,
                recipientKey,
                encryptedMsg,
                encryptedMsg.Length,
                _agentMessageParsedCallback);

            CallbackHelper.CheckResult(result);

            return taskCompletionSource.Task;
        }
    }
}
